File: androld/app/src/main/java/z3d/program/Shader3D.java
package z3d.program;


import android.opengl.GLES20;



public class Shader3D {

    public int program;


    public boolean encode() {

        this.program= uCreateGlProgram(this.getVertexShaderString(),this.getFragmentShaderString());

        return true;

    }
    public boolean  encodeVstr(String vertex,String fragment)
    {
        this.program= uCreateGlProgram(vertex,fragment);


        return true;
    }

    //创建GL程序
    public static int uCreateGlProgram(String vertexSource, String fragmentSource){



        int vertex=uLoadShader(GLES20.GL_VERTEX_SHADER,vertexSource);
        if(vertex==0)return 0;
        int fragment=uLoadShader(GLES20.GL_FRAGMENT_SHADER,fragmentSource);
        if(fragment==0)return 0;
        int program= GLES20.glCreateProgram();
        if(program!=0){
            GLES20.glAttachShader(program,vertex);
            GLES20.glAttachShader(program,fragment);
            GLES20.glLinkProgram(program);

        }
        return program;
    }

    //加载shader
    public static int uLoadShader(int shaderType, String source){
        int shader= GLES20.glCreateShader(shaderType);
        if(0!=shader){
            GLES20.glShaderSource(shader,source);
            GLES20.glCompileShader(shader);
        }
        return shader;
    }


    public String getVertexShaderString() {
        return "";
    }
    public String getFragmentShaderString() {
        return "";
    }
    public void destory() {

    }
}

File: iphone/iosgl/iosgl/bt/program/Shader3D.m
//
//  Shader3D.m
//  iosgl
//
//  Created by zhao on 7/12/2019.
//  Copyright © 2019 zhao. All rights reserved.
//
#import <GLKit/GLKit.h>
#import "Shader3D.h"

@implementation Shader3D
 
-(void)encodeVstr:(NSString*)vstr encodeFstr:(NSString*)fstr;
{
    
    if(!vstr){
        vstr=[self getVertexShaderString];
    }
    if(!fstr){
        fstr=[self getFragmentShaderString];
    }
    
    GLuint verShader,fragShader;
    _program = glCreateProgram();
    [self compileShaderStr:&verShader type:GL_VERTEX_SHADER str:vstr];
    [self compileShaderStr:&fragShader type:GL_FRAGMENT_SHADER str:fstr];
    
    glAttachShader(_program, verShader);
    glAttachShader(_program, fragShader);
    glDeleteShader(verShader);
    glDeleteShader(fragShader);
    
    //5、链接
    glLinkProgram(_program);
    
    GLint linkStatus;
    glGetProgramiv(_program, GL_LINK_STATUS, &linkStatus);
    if(linkStatus==GL_FALSE)
    {
        //获取失败信息
        GLchar message[512];
        //来检查是否有error，并输出信息
        /*
         作用:连接着色器程序也可能出现错误，我们需要进行查询，获取错误日志信息
         参数1: program 着色器程序标识
         参数2: bufsize 最大日志长度
         参数3: length 返回日志信息的长度
         参数4：infoLog 保存在缓冲区中
         */
        glGetProgramInfoLog(_program, sizeof(message), 0, &message[0]);
        
        //将C语言字符串转换成OC字符串
        NSString * messageStr = [NSString stringWithUTF8String:message];
        
        NSLog(@"Program Link Error:%@",messageStr);
        
        return;
    }
}
-(NSString*)vertexStr;
{
    if(self.vertex){
        return self.vertex;
    }else{
        return [self getVertexShaderString];
    }
}
-(void)compileShaderStr:(GLuint *)shader type:(GLenum)type str:(NSString *)str;
{
    const GLchar * source = (GLchar *)[str UTF8String];
    *shader = glCreateShader(type);
    glShaderSource(*shader, 1, &source,NULL);
    glCompileShader(*shader);
    
    
}
-(NSString *)getVertexShaderString;{
    return   @"";
}
-(NSString *)getFragmentShaderString;{
    return  @"";
}
@end
